internal/pkg/agent/application/endpoint_component_modifier.go

```diff
@@ -0,0 +1,263 @@
+		unit := comps[compIdx].Units[unitIdx]
+		unitCfgMap := unit.Config.Source.AsMap()
+
+		// ensure the following config exists:
+		// fleet.ssl:
+		//   key_passphrase_path
+		//   certificate
+		//   key
+		fleetNode, ok := unitCfgMap["fleet"]
+		if !ok {
+			// if 'fleet' isn't, present nothing to do
+			return comps, nil
+		}
+		fleetMap, ok := fleetNode.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("EndpointTLSComponentModifier: 'fleet' node isn't a map, it is: %T", fleetNode)
+		}
+
+		sslNode, ok := fleetMap["ssl"]
+		if !ok {
+			// 'ssl' node not present isn't an issue
+			return comps, nil
+		}
+		sslMap, ok := sslNode.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("EndpointTLSComponentModifier: 'ssl' node isn't a map, it is: %T", sslNode)
+		}
+
+		keyPassPathI, ok := sslMap["key_passphrase_path"]
+		if !ok {
+			// if no key_passphrase_path, nothing to decrypt
+			return comps, nil
+		}
+		keyPassPath, ok := keyPassPathI.(string)
+		if !ok {
+			return nil, errors.New("EndpointTLSComponentModifier: 'key_passphrase_path' isn't a string")
+		}
+		if keyPassPath == "" {
+			// key_passphrase_path shouldn't be empty, but if it's, nothing to decrypt
+			return comps, nil
+		}
+
+		keyI, ok := sslMap["key"]
+		if !ok {
+			// if there is a key_passphrase_path, the key must be present
+			return nil, errors.New("EndpointTLSComponentModifier: 'key_passphrase_path' present, but 'key' isn't present")
+		}
+		keyPath, ok := keyI.(string)
+		if !ok {
+			return nil, fmt.Errorf("EndpointTLSComponentModifier: 'key' isn't a string, it's %T", keyI)
+		}
+
+		certI, ok := sslMap["certificate"]
+		if !ok {
+			// if there is a key_passphrase_path, the certificate must be present
+			return nil, errors.New("EndpointTLSComponentModifier: 'key_passphrase_path' present, but 'certificate' isn't present")
+		}
+		certPath, ok := certI.(string)
+		if !ok {
+			return nil, errors.New("EndpointTLSComponentModifier: 'certificate' isn't a string")
+		}
+
+		cert, key, err := loadCertificatesWithCache(log, cache, keyPassPath, certPath, keyPath)
+		if err != nil {
+			return nil, err
+		}
+
+		// remove 'key_passphrase_path' as the certificate key isn't encrypted
+		// anymore.
+		delete(sslMap, "key_passphrase_path")
+
+		// update the certificate and its key with their decrypted version.
+		sslMap["certificate"] = cert
+		sslMap["key"] = key
+
+		unitCfg, err := component.ExpectedConfig(unitCfgMap)
+		if err != nil {
+			return nil, fmt.Errorf("EndpointTLSComponentModifier: could not covert modified config to expected config: %w", err)
+		}
+
+		unit.Config = unitCfg
+		comps[compIdx].Units[unitIdx] = unit
+
+		return comps, nil
+	}
+}
+
+func loadCertificatesWithCache(log *logger.Logger, cache *tlsCache, keyPassPath string, certPath string, keyPath string) (string, string, error) {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	// cache hit
+	// hot reload of TLS files isn't supported, thus using the path as key is
+	// fine.
```

Comment:
How is this not supported? Isn't it possible for the certificate and key to come from Fleet where those values can be changed?

Even if it is the case that they con't change, would it be difficult here to make this just a little bit smarter? Like if it does change it does the correct thing. That would future proof this code path so if they can be updated to be changed this code path doesn't also need to be updated.

Reply:
> How is this not supported? Isn't it possible for the certificate and key to come from Fleet where those values can be changed?

I think you misunderstood it. What I'm saying to be unsupported is hot reload of the files, changing the file on disk and without restarting the agent, the agent pickup the new file content. Given that isn't supported, using the file path as cache key is fine.

If fleet-server sends a new config, a different file path, then it'd be a cache miss, the new file would be read, decrypted and added to the cache.

Reply:
@blakerouse, so, what do u think?

Reply:
@AndersonQ can you point to the code where if the file path changes that it would result in it being used? What I see from the code if the passcode file path doesn't change then it will not be read again at all.

https://github.com/elastic/elastic-agent/pull/5542/files#diff-10a613fe5fe9a1e7985f230c74d11378c33f842e882e21bb3af5be2bea680e53R192

Maybe I am miss-understanding the flow here, but to me that means it will just check that and move on.